File: clients/windows/Mailcal.Tests/InvitationFormatTests.cs
// The invitation card's rules and its preview's geometry.
//
// Everything here is something a screenshot would not catch: whether "we haven't looked" reads
// differently from "nothing", whether the attendee buckets add up, whether the preview's hour band
// actually contains the meeting. The card itself is a WinUI surface, but none of these rules are,
// which is the whole reason they live in plain functions (InvitationFormat), and the reason those
// functions return a *choice* rather than a localised string: L10n.cs cannot be linked into this
// assembly, so a rule phrased as a string would be a rule no test could reach. The words are
// InvitationText's, one switch away, and that switch is exhaustive over these enums.
//
// The twin of Android's InvitationFormatTest.kt and Apple's InvitationFormatTests.swift.
using System.Globalization;
using Allodia.Mailcal.Calendar;
using uniffi.mailcal_bindings;
using Xunit;

namespace Allodia.Mailcal.Tests;

public class InvitationFormatTests
{
    // en-GB, so a day-month order and a 24-hour clock are the baseline the assertions read against.
    private static readonly CultureInfo En = new("en-GB");

    private static AttendeeTally Tally(
        uint total, uint accepted = 0, uint declined = 0, uint tentative = 0, uint needsAction = 0) =>
        new(total, accepted, declined, tentative, needsAction);

    // ---- The conflict line -------------------------------------------------------------------

    [Fact]
    public void None_one_and_many_are_three_different_sentences()
    {
        // "0 other things" and "1 other things" are not sentences, so each gets its own wording.
        Assert.Equal(ConflictLine.None, InvitationFormat.Conflicts(0, known: true));
        Assert.Equal(ConflictLine.One, InvitationFormat.Conflicts(1, known: true));
        Assert.Equal(ConflictLine.Many, InvitationFormat.Conflicts(3, known: true));
    }

    [Fact]
    public void An_unread_calendar_does_not_claim_the_day_is_free()
    {
        // The failure this guards actually shipped: mail syncs before calendars, so an invitation
        // opened on a cold start reached the card builder with nothing expanded, and the card said
        // "Nothing else in your calendar then" over a Monday holding two meetings.
        Assert.Equal(ConflictLine.Unknown, InvitationFormat.Conflicts(0, known: false));
        // And the count is not merely unprinted, it is not consulted at all.
        Assert.Equal(ConflictLine.Unknown, InvitationFormat.Conflicts(7, known: false));
    }

    // ---- The attendee tally ------------------------------------------------------------------

    [Fact]
    public void An_invitation_with_only_you_on_it_says_so()
    {
        // "1 of 1 accepted" is arithmetic, not a sentence about a meeting.
        Assert.Equal(
            new[] { AttendeeLine.OnlyYou },
            InvitationFormat.Attendees(Tally(total: 1, needsAction: 1)));
    }

    [Fact]
    public void An_invitation_with_no_attendees_says_nothing()
    {
        Assert.Empty(InvitationFormat.Attendees(Tally(total: 0)));
    }

    [Fact]
    public void Every_non_empty_bucket_earns_a_phrase()
    {
        // The four buckets sum to the total, so a line that drops one reads as arithmetic that does
        // not add up, the user counts three names in a five-person meeting and distrusts the rest.
        Assert.Equal(
            new[]
            {
                AttendeeLine.AcceptedOfTotal,
                AttendeeLine.TentativeOne,
                AttendeeLine.DeclinedOne,
                AttendeeLine.PendingOne,
            },
            InvitationFormat.Attendees(
                Tally(total: 5, accepted: 2, declined: 1, tentative: 1, needsAction: 1)));
    }

    [Fact]
    public void An_empty_bucket_is_left_out()
    {
        Assert.Equal(
            new[] { AttendeeLine.AcceptedOfTotal },
            InvitationFormat.Attendees(Tally(total: 3, accepted: 3)));
    }

    [Fact]
    public void One_person_is_not_described_in_the_plural()
    {
        // Dutch needs a different verb at one, "1 moeten nog antwoorden" is wrong, and the catalog
        // has no plural machinery, so each count-of-one is its own string. English reads fine either
        // way, which is exactly why this shipped unnoticed until the card was read in Dutch.
        Assert.Equal(
            new[]
            {
                AttendeeLine.AcceptedOfTotal,
                AttendeeLine.Tentative,
                AttendeeLine.Declined,
                AttendeeLine.Pending,
            },
            InvitationFormat.Attendees(
                Tally(total: 9, accepted: 1, declined: 2, tentative: 2, needsAction: 4)));
    }

    // ---- "When" ------------------------------------------------------------------------------

    [Fact]
    public void A_timed_meeting_names_the_day_once()
    {
        var line = InvitationFormat.When(
            "2026-01-19T09:30:00Z", "2026-01-19T10:30:00Z",
            allDay: false, zone: "UTC", use24Hour: true, culture: En);
        Assert.Contains("09:30", line, StringComparison.Ordinal);
        Assert.Contains("10:30", line, StringComparison.Ordinal);
        // One date, not two, start and end share a day.
        Assert.Equal(1, line.Split("2026").Length - 1);
    }

    [Fact]
    public void The_clock_follows_the_users_setting_not_the_culture()
    {
        // Mail and calendar must not disagree with each other, so this reads the app's own 12/24-hour
        // preference rather than what en-GB happens to default to.
        var twelve = InvitationFormat.When(
            "2026-01-19T14:00:00Z", "2026-01-19T15:00:00Z",
            allDay: false, zone: "UTC", use24Hour: false, culture: En);
        Assert.Contains("2:00", twelve, StringComparison.Ordinal);
        Assert.DoesNotContain("14:00", twelve, StringComparison.Ordinal);
    }

    [Fact]
    public void A_one_day_all_day_event_reads_as_one_date()
    {
        // The stored end is EXCLUSIVE: a single all-day event ends at the next midnight, and naming
        // both would tell the user it lasts two days.
        var line = InvitationFormat.When(
            "2026-01-19T00:00:00Z", "2026-01-20T00:00:00Z",
            allDay: true, zone: "UTC", use24Hour: true, culture: En);
        Assert.DoesNotContain("–", line, StringComparison.Ordinal);
    }

    [Fact]
    public void A_multi_day_all_day_event_names_its_inclusive_last_day()
    {
        var line = InvitationFormat.When(
            "2026-01-19T00:00:00Z", "2026-01-22T00:00:00Z",
            allDay: true, zone: "UTC", use24Hour: true, culture: En);
        Assert.Contains("–", line, StringComparison.Ordinal);
        Assert.Contains("21", line, StringComparison.Ordinal);
        Assert.DoesNotContain("22", line, StringComparison.Ordinal);
    }

    [Fact]
    public void The_display_zone_moves_the_clock()
    {
        var line = InvitationFormat.When(
            "2026-01-19T09:30:00Z", "2026-01-19T10:30:00Z",
            allDay: false, zone: "Europe/Amsterdam", use24Hour: true, culture: En);
        Assert.Contains("10:30", line, StringComparison.Ordinal);
        Assert.Contains("11:30", line, StringComparison.Ordinal);
    }

    [Fact]
    public void An_unparseable_instant_yields_no_line_rather_than_a_wrong_one()
    {
        Assert.Equal(string.Empty, InvitationFormat.When("", "", false, "UTC", true, En));
        Assert.Equal(string.Empty, InvitationFormat.When("not-a-date", "", false, "UTC", true, En));
    }

    [Fact]
    public void An_unresolvable_zone_still_draws_the_card()
    {
        // The core names an IANA zone; a host that cannot resolve one must show the meeting in the
        // wrong zone *visibly* rather than blank the "when" row and say nothing at all.
        Assert.NotEqual(
            string.Empty,
            InvitationFormat.When(
                "2026-01-19T09:30:00Z", "2026-01-19T10:30:00Z",
                allDay: false, zone: "Mars/Olympus_Mons", use24Hour: true, culture: En));
    }

    // ---- The preview's hour band -------------------------------------------------------------

    [Fact]
    public void The_span_always_contains_the_meeting()
    {
        var span = InvitationFormat.PreviewSpan(new MinuteSpan(13 * 60, 14 * 60), []);
        Assert.True(span.First <= 13);
        Assert.True(span.Last >= 14);
    }

    [Fact]
    public void A_short_meeting_on_an_empty_day_still_gets_context_around_it()
    {
        // A 30-minute meeting padded an hour each side is a two-hour sliver with nothing to compare
        // it against; the floor is what makes the picture worth drawing.
        var span = InvitationFormat.PreviewSpan(new MinuteSpan(10 * 60, (10 * 60) + 30), []);
        Assert.True(span.Count >= 6, $"{span.First}..{span.Last}");
    }

    [Fact]
    public void The_band_keeps_the_meeting_away_from_its_edges()
    {
        // Padding grown alternately, not all onto one end: a meeting pinned to the top of its own
        // preview reads as if the day starts there.
        var span = InvitationFormat.PreviewSpan(new MinuteSpan(14 * 60, 15 * 60), []);
        Assert.True(span.First < 14, $"{span.First}");
        Assert.True(span.Last > 16, $"{span.Last}");
    }

    [Fact]
    public void A_block_ending_mid_hour_keeps_the_whole_hour()
    {
        // Ceil, not truncate: an event ending 09:15 whose hour was floored would be drawn past the
        // bottom edge of its own preview.
        var span = InvitationFormat.PreviewSpan(new MinuteSpan(8 * 60, (9 * 60) + 15), []);
        Assert.True(span.Last >= 10, $"{span.Last}");
    }

    [Fact]
    public void The_band_covers_every_clash_in_full()
    {
        // The one thing that may not fall outside it. A conflict is by definition an event
        // overlapping the meeting, and it has to be drawn *whole*, a long booking cut off at the
        // top edge loses its title with it, which is exactly what the band exists to show.
        var span = InvitationFormat.PreviewSpan(
            new MinuteSpan(14 * 60, 15 * 60),
            [new MinuteSpan(9 * 60, 16 * 60)]);
        Assert.True(span.First <= 9, $"{span.First}");
        Assert.True(span.Last >= 16, $"{span.Last}");
    }

    [Fact]
    public void The_band_leaves_out_the_rest_of_the_day()
    {
        // …and everything that does NOT clash is left out, which is what buys the hours their
        // height. The card states the count in words above the grid and the disclosure label says
        // "around this meeting", so nothing is hidden without saying so.
        var span = InvitationFormat.PreviewSpan(
            new MinuteSpan(14 * 60, 15 * 60),
            [new MinuteSpan(8 * 60, 9 * 60), new MinuteSpan(21 * 60, 22 * 60)]);
        Assert.True(span.First > 8, $"{span.First}");
        Assert.True(span.Last <= 21, $"{span.Last}");
    }

    [Fact]
    public void A_block_ending_as_the_meeting_begins_is_not_a_clash()
    {
        // Half-open on both sides, exactly as the core's conflict count overlaps: back-to-back is
        // how a diary is packed, and widening the band for it would undo the zoom on every meeting
        // that follows another.
        var span = InvitationFormat.PreviewSpan(
            new MinuteSpan(14 * 60, 15 * 60),
            [new MinuteSpan(6 * 60, 14 * 60)]);
        Assert.True(span.First > 6, $"{span.First}");
    }

    [Fact]
    public void The_span_never_leaves_the_day()
    {
        var span = InvitationFormat.PreviewSpan(new MinuteSpan(23 * 60, 24 * 60), []);
        Assert.True(span.First >= 0);
        Assert.True(span.Last <= 24);
    }

    [Fact]
    public void A_squeezed_span_labels_fewer_hours_rather_than_overlapping_them()
    {
        Assert.Equal(1, InvitationFormat.PreviewStride(40));
        Assert.True(InvitationFormat.PreviewStride(5) > 1);
        // A degenerate height must not divide by zero or return a stride of 0, the ruler takes a
        // modulo by it, so zero is an exception and not merely an ugly picture.
        Assert.Equal(1, InvitationFormat.PreviewStride(0));
        Assert.Equal(1, InvitationFormat.PreviewStride(-3));
    }

    // ---- The preview's height -----------------------------------------------------------------

    [Fact]
    public void Every_band_the_span_can_produce_can_title_a_one_hour_block()
    {
        // The one thing the preview has to say. The band and the box are two halves of one rule,
        // narrow the band, or grow the box, and only their *ratio* decides whether a block gets a
        // title. So compose them, rather than pinning either number.
        //
        // That composition is the whole reason MinimumTitledHeight moved out of the WinUI view and
        // into InvitationFormat: this assembly cannot link Views/.
        for (var hours = 6; hours <= 12; hours++)
        {
            var hourHeight = InvitationFormat.PreviewHeight(hours) / hours;
            Assert.True(
                hourHeight >= InvitationFormat.MinimumTitledHeight,
                $"a one-hour block must carry its title over a {hours}-hour band, got {hourHeight}");
        }
    }

    [Fact]
    public void The_box_only_grows_when_the_band_cannot_stay_narrow()
    {
        // The ordinary case is the plain height: the band is six hours, so there is nothing to fix.
        Assert.Equal(132, InvitationFormat.PreviewHeight(6));
        // A long booking the meeting sits inside forces a wider band; the box follows it…
        Assert.True(InvitationFormat.PreviewHeight(10) > InvitationFormat.PreviewHeight(6));
        // …but stops, rather than pushing the message itself off the screen.
        Assert.Equal(240, InvitationFormat.PreviewHeight(24));
    }

    // ---- The meeting's own window ------------------------------------------------------------

    [Fact]
    public void The_meeting_window_is_wall_clock_minutes_in_the_layout_zone()
    {
        Assert.Equal(
            new MinuteSpan((10 * 60) + 30, (11 * 60) + 30),
            InvitationFormat.MeetingMinuteSpan(
                "2026-01-19T09:30:00Z", "2026-01-19T10:30:00Z", "Europe/Amsterdam"));
    }

    [Fact]
    public void A_meeting_running_past_midnight_ends_at_the_bottom_of_its_day()
    {
        var span = InvitationFormat.MeetingMinuteSpan(
            "2026-01-19T22:00:00Z", "2026-01-20T01:00:00Z", "UTC");
        Assert.Equal(22 * 60, span.Start);
        Assert.Equal(24 * 60, span.End);
    }

    [Fact]
    public void An_unparseable_instant_still_draws_the_day_it_was_given()
    {
        Assert.Equal(
            new MinuteSpan(0, 60), InvitationFormat.MeetingMinuteSpan("not-a-date", "", "UTC"));
    }

    // ---- The hold, and the write -------------------------------------------------------------

    [Fact]
    public void Only_an_unanswered_invitation_is_a_hold()
    {
        Assert.True(InvitationFormat.IsAwaitingResponse(ResponseStatus.NeedsAction));
        Assert.False(InvitationFormat.IsAwaitingResponse(ResponseStatus.Accepted));
        Assert.False(InvitationFormat.IsAwaitingResponse(ResponseStatus.Tentative));
        Assert.False(InvitationFormat.IsAwaitingResponse(ResponseStatus.Delegated));
        // Declined never reaches a client, the core hides those from every calendar surface, but
        // if one ever did it is not a hold either.
        Assert.False(InvitationFormat.IsAwaitingResponse(ResponseStatus.Declined));
    }

    [Fact]
    public void A_failed_answer_says_so_and_a_finished_one_says_nothing()
    {
        // The asymmetry is the whole rule. Once the answer lands the card already shows it, it
        // re-reads the calendar, so a second "sent" line is noise. A *failure* is the one that must
        // never be silent: the card would otherwise sit there showing the previous answer while the
        // organiser heard nothing, which is exactly the outcome this feature exists to prevent.
        Assert.Equal(WriteLine.None, InvitationFormat.Write(CalendarWriteStatus.Saved));
        Assert.Equal(WriteLine.None, InvitationFormat.Write(CalendarWriteStatus.Idle));
        Assert.Equal(WriteLine.Sending, InvitationFormat.Write(CalendarWriteStatus.Saving));
        Assert.Equal(WriteLine.Failed, InvitationFormat.Write(CalendarWriteStatus.Failed));
    }

    [Fact]
    public void A_hold_is_faded_and_a_commitment_is_untouched()
    {
        // The fade is what makes a hold read as provisional beside a booked commitment. It is applied
        // when the page is built, never inside a frame (§7), this is the arithmetic that does it.
        Assert.Equal(255, InvitationFormat.HoldAlpha(255, awaiting: false));
        Assert.Equal(102, InvitationFormat.HoldAlpha(255, awaiting: true));
        // Already-faded input fades again from where it is, rather than snapping to a fixed value.
        Assert.Equal(40, InvitationFormat.HoldAlpha(100, awaiting: true));
        // A colour that is already invisible cannot be made more so, and must not wrap round a byte.
        Assert.Equal(0, InvitationFormat.HoldAlpha(0, awaiting: true));
        Assert.Equal(1, InvitationFormat.HoldAlpha(2, awaiting: true));
    }

    [Fact]
    public void A_hold_says_so_in_the_spoken_label_and_a_commitment_does_not()
    {
        // This is the half of the treatment that is NOT optional. A dashed border and a hatched
        // gutter are invisible to a screen reader, so a grid that only draws the hold tells a
        // screen-reader user that an unanswered invitation is a settled commitment
        // (docs/calendar.md §4). Every surface that can show a hold runs through here.
        const string label = "Standup, 09:00-09:15, Work";
        const string awaiting = "Awaiting your response";

        Assert.Equal(
            "Standup, 09:00-09:15, Work, Awaiting your response",
            InvitationFormat.SpokenWithHold(label, awaiting, ResponseStatus.NeedsAction));
        Assert.Equal(label, InvitationFormat.SpokenWithHold(label, awaiting, ResponseStatus.Accepted));
        Assert.Equal(label, InvitationFormat.SpokenWithHold(label, awaiting, ResponseStatus.Tentative));
        Assert.Equal(label, InvitationFormat.SpokenWithHold(label, awaiting, ResponseStatus.Delegated));
    }
}

File: clients/apple/Packages/MailcalKit/Tests/MailcalUITests/InvitationFormatTests.swift
// The invitation card's own rules, pinned as plain values (see InvitationFormat.swift).
//
// These are the parts a screenshot cannot check: that "0 other things" and "1 other things" never get
// printed, that the four attendee buckets all reach the line so the arithmetic adds up, that an
// all-day meeting's exclusive end reads as one date rather than two, and that the preview's hour span
// contains every block on the day, because a preview that clips is a preview that lies.
//
// Deliberately locale-pinned where a string is asserted: the point is *which* wording the rule picks,
// not how Dutch spells July (the Android suite's own trap, recorded in AGENTS.md).

import Foundation
import MailcalBindings
import SwiftUI
import Testing

@testable import MailcalUI

@Suite struct InvitationFormatTests {

    private func tally(
        total: UInt32,
        accepted: UInt32 = 0,
        declined: UInt32 = 0,
        tentative: UInt32 = 0,
        needsAction: UInt32 = 0
    ) -> AttendeeTally {
        AttendeeTally(
            total: total,
            accepted: accepted,
            declined: declined,
            tentative: tentative,
            needsAction: needsAction
        )
    }

    // MARK: - Conflicts, in words

    @Test func conflictWordingHasItsOwnZeroAndOne() {
        // The three cases are distinct strings, not one template with a number in it: "0 other
        // things in your calendar" and "1 other things" are not sentences.
        let none = invitationConflictLine(count: 0, known: true)
        let one = invitationConflictLine(count: 1, known: true)
        let many = invitationConflictLine(count: 4, known: true)
        #expect(none != one)
        #expect(one != many)
        #expect(!none.contains("0"))
        #expect(many.contains("4"))
    }

    @Test func anUnreadCalendarDoesNotClaimTheDayIsFree() {
        // The regression this exists for: opening an invitation before the calendar had synced said
        // "Nothing else in your calendar then" over a diary nobody had read, and the day in fact had
        // two clashes on it. Unknown is a THIRD sentence, never the zero one.
        let unknown = invitationConflictLine(count: 0, known: false)
        #expect(unknown != invitationConflictLine(count: 0, known: true))
        // A count that arrived anyway must not leak into the wording.
        #expect(invitationConflictLine(count: 7, known: false) == unknown)
        #expect(!unknown.contains("7"))
    }

    // MARK: - The attendee tally

    @Test func anInvitationWithOnlyMeSaysSo() {
        // "1 of 1 accepted" is technically true and reads as a committee of one.
        #expect(invitationAttendeeLines(tally(total: 1, accepted: 1)) == [L10n.invitation_attendees_one()])
    }

    @Test func noAttendeesAtAllProducesNoLine() {
        #expect(invitationAttendeeLines(tally(total: 0)).isEmpty)
    }

    @Test func everyNonZeroBucketReachesTheLine() {
        // The four buckets sum to the total, so a bucket left out is arithmetic that does not add up.
        let lines = invitationAttendeeLines(
            tally(total: 6, accepted: 2, declined: 1, tentative: 1, needsAction: 2)
        )
        #expect(lines.count == 4)
        #expect(lines[0].contains("2") && lines[0].contains("6"))
        #expect(lines[1].contains("1"))
        #expect(lines[2].contains("1"))
        #expect(lines[3].contains("2"))
    }

    @Test func emptyBucketsAreLeftOut() {
        // Nobody has declined, so the card does not say "0 declined".
        let lines = invitationAttendeeLines(tally(total: 3, accepted: 3))
        #expect(lines.count == 1)
    }

    // MARK: - The "when" line

    @Test func aTimedMeetingNamesItsDateOnce() {
        let line = invitationWhen(
            startsAt: "2026-07-30T14:30:00Z",
            endsAt: "2026-07-30T15:00:00Z",
            allDay: false,
            zone: "Europe/Amsterdam",
            use24Hour: true,
            locale: Locale(identifier: "en_GB")
        )
        // 14:30 UTC is 16:30 in Amsterdam in July, the zone is applied, not ignored.
        #expect(line.contains("16:30"))
        #expect(line.contains("17:00"))
        #expect(line.contains("2026"))
        // One date for a meeting inside one day.
        #expect(line.filter { $0 == "–" }.count == 1)
    }

    @Test func theTwelveHourSettingIsHonoured() {
        // The user's 12/24-hour setting, not the locale's default: mail and calendar must agree.
        let line = invitationWhen(
            startsAt: "2026-07-30T14:30:00Z",
            endsAt: "2026-07-30T15:00:00Z",
            allDay: false,
            zone: "Europe/Amsterdam",
            use24Hour: false,
            locale: Locale(identifier: "en_GB")
        )
        #expect(line.contains("4:30 PM"))
        #expect(!line.contains("16:30"))
    }

    @Test func aOneDayAllDayMeetingReadsAsOneDate() {
        // The stored end is EXCLUSIVE, next midnight, so the inclusive last day is the start day.
        let line = invitationWhen(
            startsAt: "2026-07-29T22:00:00Z",
            endsAt: "2026-07-30T22:00:00Z",
            allDay: true,
            zone: "Europe/Amsterdam",
            use24Hour: true,
            locale: Locale(identifier: "en_GB")
        )
        #expect(!line.contains("–"))
        #expect(line.contains("30"))
    }

    @Test func aMultiDayAllDayMeetingNamesBothEnds() {
        let line = invitationWhen(
            startsAt: "2026-07-29T22:00:00Z",
            endsAt: "2026-08-01T22:00:00Z",
            allDay: true,
            zone: "Europe/Amsterdam",
            use24Hour: true,
            locale: Locale(identifier: "en_GB")
        )
        #expect(line.contains("–"))
        // 30 July through 1 August inclusive, never the exclusive 2nd.
        #expect(line.contains("30"))
        #expect(!line.contains(" 2 "))
    }

    @Test func anUnparseableInstantYieldsNoLineRatherThanAWrongOne() {
        #expect(
            invitationWhen(
                startsAt: "",
                endsAt: "",
                allDay: false,
                zone: "Europe/Amsterdam",
                use24Hour: true
            ).isEmpty
        )
    }

    // MARK: - The preview's hour span

    @Test func theSpanContainsEveryClashInFull() {
        // The one thing that may not fall outside the band. A conflict is by definition an event
        // overlapping the meeting, and it has to be drawn *whole*, a long booking cut off at the
        // top edge loses its title with it, which is exactly what the band exists to show.
        let span = invitationPreviewSpan(
            meeting: MinuteSpan(start: 14 * 60, end: 15 * 60),
            others: [MinuteSpan(start: 9 * 60, end: 16 * 60)]
        )
        #expect(span.lowerBound <= 9)
        #expect(span.upperBound >= 16)
    }

    @Test func theSpanLeavesOutTheRestOfTheDay() {
        // …and everything that does NOT clash is left out, which is what buys the hours their
        // height. The card states the count in words above the grid and the disclosure label says
        // "around this meeting", so nothing is hidden without saying so.
        let span = invitationPreviewSpan(
            meeting: MinuteSpan(start: 14 * 60, end: 15 * 60),
            others: [MinuteSpan(start: 8 * 60, end: 9 * 60), MinuteSpan(start: 21 * 60, end: 22 * 60)]
        )
        #expect(!span.contains(8))
        #expect(!span.contains(21))
        #expect(span.contains(14))
    }

    @Test func aBlockEndingAsTheMeetingBeginsIsNotAClash() {
        // Half-open on both sides, exactly as the core's conflict count overlaps: back-to-back is
        // how a diary is packed, and widening the band for it would undo the zoom on every meeting
        // that follows another.
        let span = invitationPreviewSpan(
            meeting: MinuteSpan(start: 14 * 60, end: 15 * 60),
            others: [MinuteSpan(start: 6 * 60, end: 14 * 60)]
        )
        #expect(!span.contains(6))
    }

    @Test func theMeetingAloneStillGetsAReadableSpan() {
        // A 30-minute meeting on an empty day would otherwise be a two-hour sliver with no context.
        let span = invitationPreviewSpan(meeting: MinuteSpan(start: 600, end: 630), others: [])
        #expect(span.count >= 6)
        #expect(span.contains(10))
    }

    @Test func theBandKeepsTheMeetingAwayFromItsEdges() {
        // Padding grown alternately, not all onto one end: a meeting pinned to the top of its own
        // preview reads as if the day starts there.
        let span = invitationPreviewSpan(meeting: MinuteSpan(start: 14 * 60, end: 15 * 60), others: [])
        #expect(span.lowerBound < 14)
        #expect(span.upperBound > 15 + 1)
    }

    @Test func theSpanNeverLeavesTheDay() {
        // Padding and the minimum both push against the ends of the day, and the band is what the
        // preview divides its height by, an hour outside 0..<24 would place blocks off the grid.
        let firstThing = invitationPreviewSpan(meeting: MinuteSpan(start: 0, end: 30), others: [])
        #expect(firstThing.lowerBound == 0)
        #expect(firstThing.upperBound <= 24)
        let lastThing = invitationPreviewSpan(
            meeting: MinuteSpan(start: 23 * 60, end: 24 * 60),
            others: []
        )
        #expect(lastThing.lowerBound >= 0)
        #expect(lastThing.upperBound == 24)
        // Both still cover the whole minimum, growing inwards when one end has run out of room.
        #expect(firstThing.count >= 6)
        #expect(lastThing.count >= 6)
    }

    @Test func aBlockEndingMidHourKeepsThatWholeHour() {
        // Ceil, not floor: a block ending at 09:15 that lost the 09:00 hour would be drawn cut off.
        let span = invitationPreviewSpan(meeting: MinuteSpan(start: 480, end: 555), others: [])
        #expect(span.upperBound >= 10)
    }

    @Test func theRulerThinsOutWhenAnHourIsTooShortToLabel() {
        #expect(invitationPreviewStride(hourHeight: 25) == 1)
        #expect(invitationPreviewStride(hourHeight: 18) == 1)
        // A 24-hour span squeezed into the preview's height: label every third hour, not every one.
        #expect(invitationPreviewStride(hourHeight: 150 / 24) > 1)
    }

    // MARK: - The preview's height

    @Test func everyBandTheSpanCanProduceCanNameAOneHourBlock() {
        // The one thing the preview has to say. The band and the box are two halves of one rule:
        // narrow the band, or grow the box, and only their *ratio* decides whether a block gets a
        // title. So compose them, rather than pinning either number: a change to the span, the
        // height, or the label threshold has to keep all three compatible.
        for hours in 6...12 {
            let hourHeight = invitationPreviewHeight(hours: hours) / CGFloat(hours)
            #expect(
                blockShowsLabel(minutes: 60, hourHeight: hourHeight),
                "a one-hour block must carry its title over a \(hours)-hour band"
            )
        }
    }

    @Test func theBoxOnlyGrowsWhenTheBandCannotStayNarrow() {
        // The ordinary case is the plain height: the band is six hours, so there is nothing to fix.
        #expect(invitationPreviewHeight(hours: 6) == 150)
        // A long booking the meeting sits inside forces a wider band; the box follows it…
        #expect(invitationPreviewHeight(hours: 10) > invitationPreviewHeight(hours: 6))
        // …but stops, rather than pushing the message itself off the screen.
        #expect(invitationPreviewHeight(hours: 24) == 240)
    }

    // MARK: - The meeting's own window, in the layout zone

    @Test func theMeetingWindowIsWallClockInTheLayoutZone() {
        let span = meetingMinuteSpan(
            startsAt: "2026-07-30T14:30:00Z",
            endsAt: "2026-07-30T15:00:00Z",
            zone: "Europe/Amsterdam"
        )
        #expect(span == MinuteSpan(start: 16 * 60 + 30, end: 17 * 60))
    }

    @Test func aMeetingRunningPastMidnightEndsAtTheBottomOfTheDay() {
        let span = meetingMinuteSpan(
            startsAt: "2026-07-30T21:00:00Z",
            endsAt: "2026-07-31T01:00:00Z",
            zone: "Europe/Amsterdam"
        )
        #expect(span.start == 23 * 60)
        #expect(span.end == 24 * 60)
    }

    // MARK: - The spoken label

    @Test func anUnansweredHoldSaysSoOutLoud() {
        // The dashed border and hatched gutter are invisible to a screen reader (docs/calendar.md §4).
        let answered = calendarEventLabel(
            title: "Standup",
            time: "09:00 – 09:15",
            calendar: "Work",
            participation: .accepted
        )
        let hold = calendarEventLabel(
            title: "Standup",
            time: "09:00 – 09:15",
            calendar: "Work",
            participation: .needsAction
        )
        #expect(!answered.contains(L10n.a11y_invitation_awaiting_response()))
        #expect(hold.contains(L10n.a11y_invitation_awaiting_response()))
        #expect(hold.hasPrefix(answered))
    }

    @Test func onlyAnUnansweredRecordIsDrawnAsAHold() {
        #expect(isAwaitingResponse(.needsAction))
        #expect(!isAwaitingResponse(.accepted))
        #expect(!isAwaitingResponse(.tentative))
        #expect(!isAwaitingResponse(.delegated))
        // Declined never reaches a client at all, the core hides those from every surface.
        #expect(!isAwaitingResponse(.declined))
    }

    @Test func onlyAHoldIsDashed() {
        #expect(participationStroke(.needsAction).dash == [3, 2])
        #expect(participationStroke(.accepted).dash.isEmpty)
    }

    // MARK: - Titles

    @Test func eachKindGetsItsOwnHeading() {
        // A cancellation must not read as an invitation still waiting for an answer, and an
        // out-of-date copy must not read as either.
        let titles = [
            invitationTitle(.rsvp),
            invitationTitle(.cancelled),
            invitationTitle(.informational),
            invitationTitle(.superseded),
        ]
        #expect(Set(titles).count == titles.count)
    }

    @Test func onlyASupersededCardExplainsItself() {
        // The other three either offer buttons or say plainly what they are. A superseded card
        // still *looks* answerable, so its missing buttons are the one absence needing a sentence.
        #expect(invitationNotice(.superseded) != nil)
        #expect(invitationNotice(.rsvp) == nil)
        #expect(invitationNotice(.cancelled) == nil)
        #expect(invitationNotice(.informational) == nil)
    }

    @Test func eachAnswerGetsItsOwnReplySubjectNamingTheMeeting() {
        // This one leaves the device: on an account whose calendar server does no scheduling, the
        // core emails the reply itself and this is the subject line the organiser reads. Three
        // answers sharing a subject would have them guessing which way we answered, and a subject
        // without the summary would leave them guessing which meeting.
        let subjects = [
            invitationReplySubject(.accept, "Sprint planning"),
            invitationReplySubject(.tentative, "Sprint planning"),
            invitationReplySubject(.decline, "Sprint planning"),
        ]
        #expect(Set(subjects).count == subjects.count)
        #expect(subjects.allSatisfy { $0.contains("Sprint planning") })
    }

    @Test func everyResponseStatusHasWording() {
        let lines = [
            invitationResponseLine(.needsAction),
            invitationResponseLine(.accepted),
            invitationResponseLine(.declined),
            invitationResponseLine(.tentative),
            invitationResponseLine(.delegated),
        ]
        #expect(Set(lines).count == 5)
        #expect(!lines.contains(""))
    }
}
